### ISSUE TITLE
Monorepo -> Tree Shaking Refactor: Analyze esbuild Output Code

### ISSUE THREAD

**Issue by holgerd77** [ethereumjs/ethereumjs-monorepo#3459]
When analyzing the code produced with `esbuild` respectively - to name it simpler - just scrolling through the produced code, e.g. when building the blockchain example from [here](npm run build && esbuild --bundle 08Blockchain.mjs --platform=browser --outfile=out1.mjs --format=esm) with `npm run build && esbuild --bundle 08Blockchain.mjs --platform=browser --outfile=out1.mjs --format=esm` it is totally mindblowing to see what's all in there together with the intuition/suspicion one is imminently getting that the *very very* most parts of this code will never be applied! 😋

We should definitely leverage this as well as an approach, this gives a very direct and clear feeling for the good, the bad and the ugly (for Blockchain I have e.g. discovered that the full Node.js `Buffer` code is still included there, still scratching my head about this 🤔).

To get some feeling, output code from these generated files look like this:

![grafik](https://github.com/ethereumjs/ethereumjs-monorepo/assets/931137/133fb2d4-0c59-4833-b27d-37a5cd823094)

Or another example:

![grafik](https://github.com/ethereumjs/ethereumjs-monorepo/assets/931137/75cc0eeb-40cd-47d4-8a73-12f0c767dcb0)

My first-round impression is there is *a lot* of room for improvement, also pointing to things going *beyond* the tree shaking topic.

**Comment by acolytec3** [comment]
@roninjin10 question for you on tree-shaking.  I've been doing some experiments with export maps that specifically define all of the entry points for a given module (specifically with `statemanager` as a starting point).  When I define an export map in `package.json` as something like:
```json
"exports: {
  './statemanager': {
"import": "dist/esm/stateManager.js"
},
'./rpcStateManager': {
"import": "dist/esm/rpcStateManager.js"
}
...
```
 and take out the root export, vite seems to be very successful at excluding both internal code from `statemanager` as well as it's dependency tree when I bundle statemanager in a react app, like this: 
![image](https://github.com/ethereumjs/ethereumjs-monorepo/assets/17355484/d7282c70-76a6-47b1-8c09-fad61a9a4986)

vs like this when I have a single entry point `'.': { "import": "dist/esm/index.js"}`
![image](https://github.com/ethereumjs/ethereumjs-monorepo/assets/17355484/7ac15000-3cad-49a6-9105-7b4a76bd0590)

In the second instance, only the unused code within the `statemanager` package is excluded while a whole bundle of transitive deps from the `statemanager` module that aren't used by the specific import in my react app (i.e. `simpleStateManager.js`) get bundled.

Do you like export maps as the main way of specifying entry points in modules to assist with tree-shaking?   I haven't tested with other bundlers/frameworks but this seemed like it worked really really well and maybe should be our goal with the next round of breaking releases (i.e. to remove the default entry point on all of our downstream libraries  and specify named entry points to reduce the overall bundle size for consuming apps).

**Comment by roninjin10** [comment]
### TLDR

For the most part, No, export maps do not affect tree shaking.

### Example

If I am an end user using webpack vite etc. and my bundler sees this:

```typescript
import { rpcStateManager } from '@ethereumjs/state-manager'
```

It will step into that file

```typescript
// index.js
export { rpcStateManager } from './rpcStateManager.js'
export {stateManager} from './statemanager.js'
```

and then it will immediately tree shake it

```typescript
// index.js
export { rpcStateManager } from './rpcStateManager.js'
// remove unused export
// export {stateManager} from './stateManager.js'
```

After it tree shakes it the bundle hit will be the same size as it is when you use exports to seperate.

### Exception

There are edge cases. For a UI like metamask that uses lavamoat, lavamoat doesn't support tree shaking. If a user isn't tree shaking themselves these seperate entrypoints are useful in this edge case because you essentially tree shaked for them.

### Testing this

If you want to test how it looks like tree shaking what you can do is create a new file treeShakingTest.ts

```ts
import {rpcStateManager} from './index.ts'
```

Change that bundle analyzer to use this as the entrypoint and you will see how it will behave in this situation. Pointing it at index.js is the equivelent to doing this

```ts
import * as stateManagerModule from '@ethereumjs/state-manager'
```

[Also I really like this online tool for doing this: ](https://bundlejs.com/). 

<img width="1242" alt="image" src="https://github.com/ethereumjs/ethereumjs-monorepo/assets/35039927/a8eb3f60-7cc4-49a1-9122-9cd92b7cd955">

**Comment by holgerd77** [comment]
Ok, but guess we really need to have a close look why this didn't work with Andrews experiments, if I understood correctly he more or less did the same thing (do a file with a specific import and then check the resulting bundle size).

At the moment I perceive these results as contradictory and not giving a clear direction yet. 🤔

**Comment by roninjin10** [comment]
I can look. Could be a bunch of things preventing tree shaking. 

Are side effects marked as false? https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free

**Comment by roninjin10** [comment]
Fave it a look and the reason tree shaking isn’t working in state manager package is this part of package.json

  "type": "commonjs",
  
Change this to module and it should work. Cjs doesn’t tree shake

**Comment by holgerd77** [comment]
@roninjin10 ah, yeah, that's very much on the plate/horizon anyhow to do the full switch and also go ESM for our internal setup, that's somewhat of a left-over! Guess we should put this very much of the front of the breaking release work, this was in the way of various stuff already and making things more complicated than it should be!

**Comment by acolytec3** [comment]
> Gave it a look and the reason tree shaking isn’t working in state manager package is this part of package.json
> 
>   "type": "commonjs",
>   
> Change this to module and it should work. Cjs doesn’t tree shake 

I'm not sure this is applicable. Our build process produces ESM and CJS outputs. I tried the same process I outlined above of switching statemanager to use an export map with named exports with webpack and it produced similar results to vite where it excluded whole dependency subtrees for the named exports that weren't imported by the consuming application so this feels like it will greatly enhance tree shaking for users of our libraries.

**Comment by roninjin10** [comment]
Trust me on this. I say this with 100% confidence export maps do not affect tree shaking and if it looks like they do that means tree shaking failed to happen. 

This can be tested using a library that has export maps like viem

**Comment by roninjin10** [comment]
Tested this in a few instances to prove it.  Setup here is a vite app that gets created with [create wagmi](https://wagmi.sh/cli/create-wagmi)

For zustand importing createStore from entyrpoint or vanilla are same
<img width="391" alt="image" src="https://github.com/ethereumjs/ethereumjs-monorepo/assets/35039927/9ca360bf-dfa9-4c8d-a2f5-9b725a9b6f2d">

Same is true for a couple of other libraries I tested except for viem which showed a small but minimal difference of 0.2kb that didn't get tree shaken from entrypoint import

> Our build process produces ESM and CJS outputs.

This could be true. Adding a console.log to the ESM imports only and then running the code is easiest way to test that it's properly using ESM. If it's is properly using ESM double check that "sideEffect": false is set in package.json

**Comment by roninjin10** [comment]
Testing tree shaking with Tevm I noticed ethereumjs tree shakes way differently depending on how aggressive or loose the tree shaking settings are set

Below we see that if I set tree shaking to smallest it tree shakes all JS while if I set it to `recomended` it includes almost all ethereumjs code. 

This implies that some rule that "smallest" doesn't follow is the root cause. Docs about what rules `smallest` ignores are [here](https://rollupjs.org/configuration-options/#treeshake). SideEffect: false in package.json would be my first guess

![telegram-cloud-photo-size-1-4933710584994704417-y](https://github.com/ethereumjs/ethereumjs-monorepo/assets/35039927/ad55f4ad-d3db-4655-a7a2-9b4937c7ef74)

**Comment by roninjin10** [comment]
The above code is using `@ethereumjs/util` as a sub dep of `@tevm/utils` I forgot to mention

**Comment by holgerd77** [comment]
Maybe not directly closing this, since this might still be worth to have some closer looks.

Just a technical update, since we are mostly bundling with vite/rollup now:

One can produce a readable bundle by adding `minify: false` as a `build` option to [vite.config.bundler.ts](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/evm/vite.config.bundler.ts).

**Comment by holgerd77** [comment]
Also just have created this gist just for fun 🙂:
https://gist.github.com/holgerd77/2c032488196b4afee5d976dc85ee70eb

(if someone wants to have a look without doing the above steps)
